### DISCUSSION TITLE
Naive question on building wxsqlite3 through msys2

### DISCUSSION THREAD

**Discussion by easior** [utelle/wxsqlite3#95] [Q&A]
Hi, I plan to do something with sqlite by using wxWidgets library on MS Windows. I found that wxSQLite3 is useful. However, I ran into a problem after I cloned the wxSQLite3 repository and built it. Specifically, I executed the following command lines in msys2 terminal:
```sh
$ cd build
$ mingw32-make
```
I got some errors:
```
==== Building wxsqlite3 (debug_win32) ====
wxsqlite3_version.rc
../src/wxsqlite3_version.rc:9: fatal error: when writing output to : Invalid argument
    9 |
      |
compilation terminated.
D:\Developer\msys64\mingw64\bin\windres.exe: can't open file `codes': No such file or directory
D:\Developer\msys64\mingw64\bin\windres.exe: preprocessing failed.
mingw32-make[1]: *** [wxsqlite3.make:259: obj/gcc/Win32/Debug/wxsqlite3/wxsqlite3_version.res] Error 1
mingw32-make: *** [Makefile:86: wxsqlite3] Error 2
```
I don't know how to resolve those errors.

**Comment by utelle** [comment]
Actually I have no idea why `windres` throws such an error message, especially, because there is no such file `codes` used in wxSQLite3. However, since `windres` complains about **writing** a file, it may help to instruct `windres` to use a temporary file. To do that try the following:

Adjust the makefile `wxsqlite3.make` by adding 
```
RESFLAGS = --use-temp-file
```
before or after line 21
```
RESCOMP = windres
```

**Reply by easior** [reply]
Thank you for your comment. It works partially. However, the search path for wxwidgets isn't correct even if I run the following comman
```sh
$ mingw32-make wxRootDir=D:\Developer\msys64\mingw64\include\wx-3.0 config=release_win64
```
or 
```sh
$ mingw32-make wxRootDir=/d/Developer/msys64/mingw64/include/wx-3.0 config=release_win64
```
After I read the wxsqlite3.make, I can't figure out how to find a proper way in msys2?

**Reply by utelle** [reply]
Is this path `/d/Developer/msys64/mingw64/include/wx-3.0` really the root directory of your wxWidgets installation? I have my doubts, as it looks more like the path for the wxWidgets header files. However, `wxRootDir` **must** point to the wxWidgets root folder (containing subfolders `include`, `src`, `lib` etc).

By the way, how did you install wxWidgets? Did you follow the instructions given on the webpage [wxWidgets with Msys backend](https://docs.wxwidgets.org/trunk/plat_msw_msys2.html), or did you do it differently? If the latter, how?

I have never used Msys myself, but AFAICR it emulates a Unix shell on Windows. So, if you built wxWidgets using the `configure/make` toolchain, then you should do the same for wxSQLite3.

**Reply by easior** [reply]
Following your instruction, I resolved my problem. Let me take a note here:
```sh
$ pacman -S autoconf m4 automake-wrapper mingw-w64-x86_64-dlfcn mingw-w64-x86_64-wxWidgets mingw-w64-x86_64-sqlite3
$ autoreconf
$ mkdir msbuild && cd msbuild
$ MAKE="mingw32-make" ../configure
$ mingw32-make
```
Except for some trivial warnings, the library is successfully built.

If I have any time in the coming week, I would like to make a mingw package for wxsqlite3.

**Reply by utelle** [reply]
> Except for some trivial warnings, the library is successfully built.

That's good news. You may show the warnings here. Unless they occur for the original SQLite source code, I'll try to fix them.

> If I have any time in the coming week, I would like to make a mingw package for wxsqlite3.

Of course, you are welcome to do so. Thanks.

**Reply by easior** [reply]
I made a mistake. In fact, there is still something wrong in the library. Let me generate some log file. Before I installed mingw-w64-x86_64-dlfcn, I got errors:
```sh
$ mingw32-make
  CXX      samples/minimal.o
  CC       src/sqlite3mc_amalgamation.lo
cc1.exe: warning: command-line option '-fpermissive' is valid for C++/ObjC++ but not for C
  CXX      src/wxsqlite3.lo
  CXXLD    libwxcode_mswu_wxsqlite3-3.0.la

*** Warning: linker path does not have real file for library -ldl.
*** I have the capability to make that library automatically link in when
*** you link to this library.  But I can only do this if you have a
*** shared version of the library, which you do not appear to have
*** because I did check the linker path looking for a file starting
*** with libdl but no candidates were found. (...for file magic test)
*** The inter-library dependencies that have been dropped here will be
*** automatically added whenever a program is linked with this library
*** or is declared to -dlopen it.

*** Since this library must not contain undefined symbols,
*** because either the platform does not support them or
*** it was explicitly requested with -no-undefined,
*** libtool will only create a static version of it.
  CXXLD    samples/minimal.exe
libtool: link: warning: `-no-install' is ignored for x86_64-w64-mingw32
libtool: link: warning: assuming `-no-fast-install' instead
D:/Developer/msys64/mingw64/bin/../lib/gcc/x86_64-w64-mingw32/10.3.0/../../../../x86_64-w64-mingw32/bin/ld.exe: cannot find -ldl
collect2.exe: error: ld returned 1 exit status
mingw32-make: *** [Makefile:642: samples/minimal.exe] Error 1
```
After it is installed, the library is successfully built only with the following additional warnings
```sh
libtool: link: warning: `-no-install' is ignored for x86_64-w64-mingw32
libtool: link: warning: assuming `-no-fast-install' instead
  CXX      samples/treeview/foldertree.o
  CXX      samples/treeview/projectlist.o
  CXX      samples/treeview/treeviewapp.o
  CXX      samples/treeview/treeviewsample.o
  CXXLD    samples/treeview/treeview.exe
libtool: link: warning: `-no-install' is ignored for x86_64-w64-mingw32
libtool: link: warning: assuming `-no-fast-install' instead
```
However, I did these procedures  on Windows 10 in another PC, it gave rises to many strip errors caused by wrong character set. I can't figure out why.

**Reply by easior** [reply]
For msys2 part, there is less progress to make a mingw package. Fortunately, a Homebrew formula (cf. https://raw.githubusercontent.com/easior/homebrew-wunderland/master/wxsqlite3.rb) has been made by me.

**Reply by utelle** [reply]
I have a few comments regarding the homebrew formula:

1) Homepage link
  Please use [**_https_**://utelle.github.io/wxsqlite3](https://utelle.github.io/wxsqlite3).
2) License
  The license is not pure _LGPL-3.0+_, but _LGPL-3.0+ **WITH WxWindows-exception-3.1**_. This is an important difference when using the library for a commercial application, because it allows to statically link the library without causing license issues.
3) Dependencies
  **wxSQLite3** does **_not_** depend on _SQLite3_, at least not in the sense that _SQLite3_ has to be installed as a prerequisite. **wxSQLite3** includes the source code amalgamation of [_SQLite3 Multiple Ciphers_](https://github.com/utelle/SQLite3MultipleCiphers), which in turn is based on a the patched source code amalgamation of _SQLite3_.

Regarding the linker error you experienced before installing `mingw-w64-x86_64-dlfcn` I think another _solution_ could be to simply remove the linker option `-ldl`, when building with _msys_. In fact, the autoconf/automake build files are meant to be used on non-Windows platforms and may need adjustments when used with _msys_.

Regarding the linker warning about the linker option `-no-install` it is most likely that simply replacing it by `-no-fast-install` in`Makefile.am` would be sufficient.
